Detect local traffic using interface

tssurya opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Implements the detection of local traffic using an interface parameter as outlined in the KEP: https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules/README.md#iptables---masquerade-off-cluster-traffic-to-services-by-node-ip



**Special notes for your reviewer**:

Though the KEP outlines an interface parameter to accept multiple prefixes, for now we are sticking to accepting a single string argument. In future when we change `--cluster-cidr` and `--node-cidr` to accept multiple values, we can also change `--pod-interface-name-prefix` and `--pod-bridge-interface`to accept multiple values.

**Does this PR introduce a user-facing change?**:
yes

```release-note
Adds support for "InterfaceNamePrefix" and "BridgeInterface" as arguments to --detect-local-mode option and also introduces a new optional `--pod-interface-name-prefix` and `--pod-bridge-interface` flags to kube-proxy.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
[[KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20191104-iptables-no-cluster-cidr.md#proposal](https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules/README.md#iptables---masquerade-off-cluster-traffic-to-services-by-node-ip)


```docs
Adds support for "BridgeInterface" and "InterfaceNamePrefix" as arguments to `--detect-local-mode` option and also introduces two new optional `--pod-interface-name-prefix` and `--pod-bridge-interface` flags to kube-proxy-config to avail the new local detector modes.
```
/sig network

aojea commented:
/ok-to-test

aojea commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -91,3 +91,33 @@ func (d *detectLocalByCIDR) JumpIfNotLocal(args []string, toChain string) []stri
+// This can be used when a pod interface name/prefix can be used to capture the notion of local traffic.
+func NewDetectLocalByInterface(ifaceName string) (LocalTrafficDetector, error) {
+	maxIfNameSize := 16 // linux/if.h
+	if len(ifaceName)+1 > maxIfNameSize || ifaceName == "" {
+		return nil, fmt.Errorf("invalid interface name")
+	}
```
This should be done here
https://github.com/kubernetes/kubernetes/blob/ee297b6f4d78daaf2f71c78a05203698581ca40a/pkg/proxy/apis/config/validation/validation.go#L39
and add tests in validation_test.go

I personally think we should validate that the prefix len is >= 4 (i.e "veth") or <= than (max-7) (ie. 2^7 is 128) , preventing people foot gun

@danwinship @thockin , what do you think?

tssurya replied:
@aojea : yeah it makes sense, I guess in plugins like sdn that use kube-proxy we use the function directly and don't use the validation parts via api, so I added it here. Makes sense to move the validation separately into validation.go.

wedaly replied:
quick comment about the minimum length validation: Azure CNI uses the prefix "azv", so if you decide to add a minimum length requirement could you set it >= 3 (or lower?)

tssurya replied:
yeah sure, I will talk with Antonio about this. For now I didn't add the min length part because like you are pointing out different platforms could be using different naming schemes, and in linux kernel `IFNAMSIZ` only includes an upper bound. If folks really want to put a lower bound which is not 0, maybe 2 ?

tssurya commented:
/remove-lifecycle stale

aojea commented:
/remove-lifecycle stale
@tssurya if you need some help or want to hand it over let me know, I have some people asking for some areas to help and this seems a good candidate

tssurya commented:
> /remove-lifecycle stale
> @tssurya if you need some help or want to hand it over let me know, I have some people asking for some areas to help and this seems a good candidate

oh no, sorry yea at this point I know what needs to be done here, wasn't on the priority radar. I'll pick this up and rebase.

tssurya commented:
/remove-lifecycle stale

tssurya commented:
I'll close this for now and reopen it when I rebase.

wedaly commented:
hi @tssurya, thanks for putting this together! I'm interested in using this for something I'm working on in Azure Kubernetes, and I was wondering if there was anything I could do to help move the PR forward?

I tried rebasing the branch onto master and testing it manually. Here's the rebased branch (feel free to take any of this if you want): https://github.com/kubernetes/kubernetes/compare/master...wedaly:detect-local-interface-rebased 

Is there anything I can do to help with this PR? Thanks!

tssurya commented:
> hi @tssurya, thanks for putting this together! I'm interested in using this for something I'm working on in Azure Kubernetes, and I was wondering if there was anything I could do to help move the PR forward?
> 
> I tried rebasing the branch onto master and testing it manually. Here's the rebased branch (feel free to take any of this if you want): [master...wedaly:detect-local-interface-rebased](https://github.com/kubernetes/kubernetes/compare/master...wedaly:detect-local-interface-rebased)
> 
> Is there anything I can do to help with this PR? Thanks!

HI @wedaly sorry did not see this comment. If you don't mind can I push this forward then?

tssurya committed:
Local Traffic Detector: Add two new modes

This PR introduces two new modes for detecting
local traffic in a cluster.
1) detectLocalByBridgeInterface: This takes a bridge name
as argument and decides all traffic that match on their
originating interface being that of this bridge, shall be
considered as local pod traffic.
2) detectLocalByInterfaceNamePrefix: This takes an interface prefix
name as argument and decides all traffic that match on their
originating interface names having a prefix that matches this
argument shall be considered as local pod traffic.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya committed:
Enable local traffic detection using the interface options

This commit adds the framework for the new local detection
modes BridgeInterface and InterfaceNamePrefix to work.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya committed:
Add validation for bridge-interface and interface-name-prefix

Co-authored-by: Will Daly <widaly@microsoft.com>
Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya commented:
test pull-kubernetes-e2e-kind

tssurya commented:
/test pull-kubernetes-e2e-kind

tssurya commented:
Tested this against a KIND cluster:
```
# necessary for conformance
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: ipv4
nodes:
# the control plane node
- role: control-plane
- role: worker
- role: worker
kubeadmConfigPatches:
- |
  apiVersion: kubeproxy.config.k8s.io/v1alpha1
  kind: KubeProxyConfiguration
  metadata:
    name: config
  detectLocalMode: "Interface"
  interfacePrefix: "veth"
```
IPTable Rules:
```
-A KUBE-SERVICES -d 10.96.0.10/32 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
-A KUBE-SVC-ERIFXISQEP7F7OF4 -d 10.96.0.10/32 ! -i veth+ -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-IT2ZTR26TO4XFPTO
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-SEP-ZXMNUKOKXUTL2MK2
-A KUBE-SEP-ZXMNUKOKXUTL2MK2 -s 10.244.0.3/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-MARK-MASQ
-A KUBE-SEP-ZXMNUKOKXUTL2MK2 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.244.0.3:53
```

aojea commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -317,3 +318,12 @@ func validateShowHiddenMetricsVersion(version string, fldPath *field.Path) field
+
+func validateInterfacePrefix(interfacePrefix string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	const maxIfNameSize = 16 // linux/if.h (as defined by IFNAMSIZ in kernel)
```
this will allow to match only one interface.
Also, if I set `--detect-local-mode` to localInterface and I forget to set the interfacePrefix it will match all the interfaces in the host.

@thockin @danwinship what do you think of make the validation something like:

```
if detect-mode == local-interface 
    validate interfaceName is > 3  < 14
```
minimum because `eth` == 3
maximum 14 because it leaves us with 2 free slots that is enough since names are alphanumerical

tssurya replied:
ack yeah I am ok with `>= 3 && <14`.

aojea replied:
it is better if you wait for Tim's and Dan's opinion :)

tssurya replied:
> this will allow to match only one interface. Also, if I set `--detect-local-mode` to localInterface and I forget to set the interfacePrefix it will match all the interfaces in the host.

actually that isn't true right? if I set `--detect-local-mode=Interface`, and interfacePrefix isn't set, it will default to no-op detect-local. A prefix has to be set if one has to use the interface mode for local traffic detector.

Also this brings me to the next question, should I be making the InterfacePrefix accept a string of comma separated names to allow match on multiple interface prefixes?

danwinship replied:
I don't think there's much benefit in validating the length of the interface name. If the user gets this argument wrong, it's likely to be because of a typo (`--interface-prefix veht`) not because they passed a string longer than 16 characters...

However, you should validate that if `--interface-prefix` is non-empty if and only if `--detect-local-mode` is `Interface`.

And in that case, you shouldn't have to re-validate that it's non empty in the proxy code.

tssurya replied:
ack, will change it per Dan's suggestion.

tssurya replied:
qq, will we always go to `getLocalDetector` code in kube-proxy only after we pass the validation? Asking because for the rest of the parameters like podCIDR/nodeCIDR we seem to revalidate their length again in the proxy code to ensure they are not empty.

danwinship commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -461,6 +461,18 @@ func getLocalDetector(mode proxyconfigapi.LocalMode, config *proxyconfigapi.Kube
 		return proxyutiliptables.NewDetectLocalByCIDR(nodeInfo.Spec.PodCIDR, ipt)
+	case proxyconfigapi.LocalModeInterfaceName:
+		if len(strings.TrimSpace(config.InterfaceName)) == 0 {
+			klog.InfoS("Detect-local-mode set to InterfaceName, but no ifacename is defined")
```
"but no interface-name is defined"

But also, you shouldn't log an Info message and fall back, you should just return an error. (The `ClusterCIDR` case has to have fallback because it's the default, but the others shouldn't. Though we probably don't want to belatedly fix the `NodeCIDR` case at this point.)

danwinship commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -461,6 +461,18 @@ func getLocalDetector(mode proxyconfigapi.LocalMode, config *proxyconfigapi.Kube
+		return proxyutiliptables.NewDetectLocalByInterfaceName(config.InterfaceName)
+	case proxyconfigapi.LocalModeInterfacePrefix:
+		if len(strings.TrimSpace(config.InterfacePrefix)) == 0 {
+			klog.InfoS("Detect-local-mode set to InterfacePrefix, but no ifaceprefix is defined")
```
"but no interface-prefix is defined"

danwinship commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -518,6 +530,30 @@ func getDualStackLocalDetectorTuple(mode proxyconfigapi.LocalMode, config *proxy
 			}
 		}
 		return localDetectors, err
+	case proxyconfigapi.LocalModeInterfaceName:
```
you could just do

```
case proxyconfigapi.LocalModeInterfaceName, proxyconfigapi.LocalModeInterfacePrefix:
        localDetector, err := getLocalDetector(...)
        if err == nil {
                localDetectors[0] = localDetector
                localDetectors[1] = localDetector
        }
        return localDetectors, err

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -2165,6 +2165,204 @@ COMMIT
 	assertIPTablesRulesEqual(t, expected, fp.iptablesData.String())
 }
 
+func TestOnlyLocalLoadBalancingDetectLocalByInterfacePrefix(t *testing.T) {
```
I'm not sure these tests add anything; the tests in `pkg/proxy/util/iptables/traffic_test.go` test that the new detectors generate the right rules, and the existing tests here in `proxier_test.go` test that, whatever rules the localdetector outputs, they get added to the iptables output as expected. So there's no need for extra per-localdetector tests here.

tssurya replied:
hmm you make a good point, Its true that nothing is extra in these tests expect its give the whole picture which the other two tests like you mention do anyways. I am happy to remove the duplicate tests.

danwinship commented:
/lgtm
/approve

/hold

I guess it would be good to have integration tests to verify that the iptables rules we're creating actually have the expected effect (ie, that if you use `detectLocalByInterfaceName` that `JumpIfLocal` generates a rule that actually does match traffic from that interface and does not match traffic from other interfaces; and similarly for `detectLocalByInterfacePrefix`... and similarly for `detectLocalByCIDR` for that matter). But I'm not sure if we have anything set up that would let us test that easily?

(`/hold`-ing to give others a chance to react, but if no one suggests anything then feel free to `/hold cancel` in a few days.)

tssurya commented:
Thanks so much Dan for reviewing!

>I guess a simpler possibility would be to just make sure that different prow jobs use different `--detect-local-mode` options...

hmm, let me take a look at the different CI jobs here and see if I can do this.

aojea commented:
why do we need `interfaceName` if we have `interfacePrefix`?
`interfaceName` works only for one interface, no? is that useful?

tssurya commented:
> why do we need `interfaceName` if we have `interfacePrefix`? `interfaceName` works only for one interface, no? is that useful?

dan was saying if users have eth1 and eth10, then we need to let them have a choice of using eth1 alone and not be forced to match on eth1 and eth10.

aojea commented:
> > why do we need `interfaceName` if we have `interfacePrefix`? `interfaceName` works only for one interface, no? is that useful?
> 
> dan was saying if users have eth1 and eth10, then we need to let them have a choice of using eth1 alone and not be forced to match on eth1 and eth10.

@danwinship does it mean the cluster will have one pod only? or am I reading it upside down and the interface-name detects the external traffic?

danwinship commented:
>  interfaceName works only for one interface, no? is that useful?

eg, [in the case of openshift-sdn](https://github.com/openshift/sdn/blob/master/pkg/cmd/openshift-sdn-node/proxy.go#L120), all pod traffic arrives at iptables from `tun0`, not from a veth directly. Although sdn doesn't create any other interfaces whose names start with `tun0` itself, it's possible someone else could. Likewise, anything else OVS-based (or anything linux-bridge based that doesn't use `bridge-nf-call-iptables`) could run into similar issues

aojea commented:
> > interfaceName works only for one interface, no? is that useful?
> 
> eg, [in the case of openshift-sdn](https://github.com/openshift/sdn/blob/master/pkg/cmd/openshift-sdn-node/proxy.go#L120), all pod traffic arrives at iptables from `tun0`, not from a veth directly. Although sdn doesn't create any other interfaces whose names start with `tun0` itself, it's possible someone else could. Likewise, anything else OVS-based (or anything linux-bridge based that doesn't use `bridge-nf-call-iptables`) could run into similar issues

TIL thanks for the clarification

aojea commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -90,3 +90,64 @@ func (d *detectLocalByCIDR) JumpIfNotLocal(args []string, toChain string) []stri
+	ifaceName string
+}
+
+// NewDetectLocalByInterfaceName implements the LocalTrafficDetector interface using an interface name.
+// This can be used when a pod interface name/prefix can be used to capture the notion of local traffic.
```
maybe we can document this better, right now is misleading without Dan's context https://github.com/kubernetes/kubernetes/pull/95400#issuecomment-1039185009

aojea replied:
```suggestion
// NewDetectLocalByInterfaceName implements the LocalTrafficDetector interface using an interface name.
// This can be used when an interface name/prefix can be used to capture the notion of local traffic.
```

aojea replied:
I think that just removing pod ...

tssurya replied:
ack, let me change this.

tssurya replied:
done.

aojea commented on cmd/kube-proxy/app/server_others_test.go:
```diff
@@ -450,6 +460,42 @@ func Test_getLocalDetector(t *testing.T) {
+		},
+		{
+			mode:        proxyconfigapi.LocalModeInterfacePrefix,
+			config:      &proxyconfigapi.KubeProxyConfiguration{InterfacePrefix: "1234567890123456789"}, // doesn't fail because validation is done before calling getLocalDetector
```
this is no longer true, right? there is no check for max length on validation, it only checks is not empty

tssurya replied:
yeah good catch, I missed this in all the iterations, let me remove the comment, though I'll leave the test case there as a reminder that we don't validate...

tssurya replied:
done/

aojea commented:
/lgtm
/hold cancel
Thanks (for your patience :smile: )

aojea commented:
/assign @danwinship 
for final approval

tssurya commented on cmd/kube-proxy/app/server.go:
```diff
@@ -209,6 +209,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.Float32Var(&o.config.ClientConnection.QPS, "kube-api-qps", o.config.ClientConnection.QPS, "QPS to use while talking with kubernetes apiserver")
 	fs.Var(&o.config.DetectLocalMode, "detect-local-mode", "Mode to use to detect local traffic. This parameter is ignored if a config file is specified by --config.")
+	fs.StringVar(&o.config.InterfaceName, "interface-name", o.config.InterfaceName, "The interface name in the cluster. Kube-proxy considers traffic as local if originating from an interface which matches the value. This argument should be set if DetectLocalMode is set to LocalModeInterfaceName.")
+	fs.StringVar(&o.config.InterfacePrefix, "interface-prefix", o.config.InterfacePrefix, "The interface prefix in the cluster. Kube-proxy considers traffic as local if originating from interfaces that match the given prefix. This argument should be set if DetectLocalMode is set to LocalModeInterfacePrefix.")
```
actually this is on me, the KEP: https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules/README.md#design-details says `--pod-interface-prefix`.. Should I stick to that? I also like `--detect-local-traffic-interface-name` like you are suggesting.

danwinship replied:
`--pod-interface-prefix` works for the InterfacePrefix case, but you can't really do `--pod-interface-name` for the InterfaceName case, because it's not the name of a pod interface, it's the name of the interface you go through on your way to the pod interface...

It could maybe be `--pod-bridge-interface-name` but then I might want to change the `--detect-local-mode` value to include "bridge" too?

Having a ridiculously long command line option name like `--detect-local-traffic-interface-name` is not necessarily bad because it encourages people to move away from command-line options and start using a YAML config file. :slightly_smiling_face:

tssurya replied:
So let's go with `--pod-interface-prefix` and `--pod-bridge-interface-name` and I'll change it everywhere in this PR i.e replace interfaceprefix and interfacename with these two ^ ones. Will rebase with the changes in 24hrs... Just want to give some time for you both to ack these ^ fields.

tssurya replied:
done.

danwinship replied:
> done

forgot to `git push`?

tssurya replied:
yeah done came out faster than I'd have liked. :D

wedaly commented:
> I guess it would be good to have integration tests to verify that the iptables rules we're creating actually have the expected effect (ie, that if you use `detectLocalByInterfaceName` that `JumpIfLocal` generates a rule that actually does match traffic from that interface and does not match traffic from other interfaces; and similarly for `detectLocalByInterfacePrefix`... and similarly for `detectLocalByCIDR` for that matter). But I'm not sure if we have anything set up that would let us test that easily?

@khenidak mentioned that we have some folks in Azure with experience adding these kinds of tests. We'd be happy to help out with this!

tssurya commented:
> > I guess it would be good to have integration tests to verify that the iptables rules we're creating actually have the expected effect (ie, that if you use `detectLocalByInterfaceName` that `JumpIfLocal` generates a rule that actually does match traffic from that interface and does not match traffic from other interfaces; and similarly for `detectLocalByInterfacePrefix`... and similarly for `detectLocalByCIDR` for that matter). But I'm not sure if we have anything set up that would let us test that easily?
> 
> @khenidak mentioned that we have some folks in Azure with experience adding these kinds of tests. We'd be happy to help out with this!

thanks so much @wedaly ! much appreciated! would you like to open an issue and assign it to the interested person?

aramase commented:
> thanks so much @wedaly ! much appreciated! would you like to open an issue and assign it to the interested person?

I think we can start off with opening an issue for follow-up. In terms of the test since we have different flavors with kind today (ipv6, ipv4), should we have more flavors with different `--detect-local-mode` to validate this config? We could build a test matrix to validate features that are enabled during cluster bootstrap.

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -209,6 +209,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.Float32Var(&o.config.ClientConnection.QPS, "kube-api-qps", o.config.ClientConnection.QPS, "QPS to use while talking with kubernetes apiserver")
 	fs.Var(&o.config.DetectLocalMode, "detect-local-mode", "Mode to use to detect local traffic. This parameter is ignored if a config file is specified by --config.")
+	fs.StringVar(&o.config.DetectLocal.BridgeInterfaceName, "pod-bridge-interface-name", o.config.DetectLocal.BridgeInterfaceName, "A bridge interface name in the cluster. Kube-proxy considers traffic as local if originating from an interface which matches the value. This argument should be set if DetectLocalMode is set to LocalModeBirdgeInterfaceName.")
```
s/LocalModeBirdgeInterfaceName/LocalModeBridgeInterfaceName/

aojea replied:
actually, this is not the name of the value, is the name of the constant used internally.
We should use the name the users should use 

```
	LocalModeClusterCIDR         LocalMode = "ClusterCIDR"
	LocalModeNodeCIDR            LocalMode = "NodeCIDR"
	LocalModeBridgeInterfaceName LocalMode = "BridgeInterfaceName"
	LocalModeInterfacePrefix     LocalMode = "InterfacePrefix"
```

tssurya replied:
Ah I made a mistake in the doc string . Will fix that (on PTO rn) next week. I’m not sure I follow the second comment? Since we are sub-configuring it within DetectLocal it doesn’t make sense to add Local again right ? cc @danwinship

tssurya replied:
Sorry ignore my previous comment. I get what you meant. I’m guessing the last word in the doc string needs to be changed to BridgeInterfaceName since that’s what users will set.

aojea replied:
right

tssurya replied:
done.

monubamb100 commented:
Do we know why the PR is taking time to merge?

monubamb100 commented:
Few of our customers are waiting on this PR to be merged. Do we know when it will merged?

aojea commented:
code freeze is on 2 weeks,  we are on time, is not like people is using master in production :smile: 

> 01:00 UTC Wednesday 30th March 2022 / 02:00 BST Wednesday 30th March 2022 / 18:00 PDT Tuesday 29th March 2022: Week 12

https://www.kubernetes.dev/resources/release/#summary

wedaly commented:
I ran the conformance and sig-network E2E tests on this PR in a kind cluster, setting `detectLocalMode="InterfacePrefix"` and `detectLocal.interfacePrefix="veth"`. Happy to report that all the tests I ran passed! 🎉 

With k/k commit 46693273e2120034c7049e8a7a364efde27cccf4 and https://github.com/wedaly/test-infra/commit/f412d4e8d70b39b515067a5e53942686464a366a
```
KUBE_PROXY_DETECT_LOCAL_MODE=InterfacePrefix \
FOCUS="\[sig-network\]|\[Conformance\]"  \
SKIP="\[Feature:(Networking-IPv6|Example|Federation|PerformanceDNS)\]|LB.health.check|LoadBalancer|load.balancer|GCE|NetworkPolicy|DualStack" \
../../k8s.io/test-infra/experiment/kind-detect-local-e2e.sh

# ... 2 hours later

Ran 411 of 7117 Specs in 7476.031 seconds
SUCCESS! -- 411 Passed | 0 Failed | 0 Pending | 6706 Skipped
PASS
```

tssurya commented:
oops I need to rebase on https://github.com/kubernetes/kubernetes/commit/e3549646ec30948bc489e94a2f2f69a238504ec0

aojea commented on staging/src/k8s.io/legacy-cloud-providers/azure/clients/armclient/mockarmclient/interface.go:
```diff
@@ -1,4 +1,3 @@
-//go:build !providerless
```
@tssurya this change is unrelated and makes the job fail, it seems you only need
```
pkg/generated/openapi/zz_generated.openapi.go
```

tssurya replied:
ahh yeah probably oops :)

tssurya replied:
done.

tssurya commented:
@aojea and @danwinship : PTAL, I think we are almost there? :)

khenidak commented:
I understand that there is pressure on getting this merged. And we should. But in the same time i think it is too risky moving this forward without the supporting e2e test. Even if we have to wait. I think that we are already light on e2e with regards to local traffic detection, adding that will create a situation where bugs will be very hard to track down and thus correcting.

/hold

tssurya commented:
> I understand that there is pressure on getting this merged. And we should. But in the same time i think it is too risky moving this forward without the supporting e2e test. 

is it? asking since by default these modes won't be enabled.

> Even if we have to wait. I think that we are already light on e2e with regards to local traffic detection, adding that will create a situation where bugs will be very hard to track down and thus correcting.
> 

Do we already have e2e's for the other two detect local traffic modes which are already merged upstream? I can take a look at the sample and try to do the same.

fedebongio commented:
/remove-sig api-machinery

aojea commented:
> I understand that there is pressure on getting this merged. And we should. But in the same time i think it is too risky moving this forward without the supporting e2e test. Even if we have to wait. I think that we are already light on e2e with regards to local traffic detection, adding that will create a situation where bugs will be very hard to track down and thus correcting.

There is a job that will run periodically using this detector, there is no more room for presubmits :/ , but this change is not very risky and @wedaly tested it. 
By the way, @wedaly can you please submit a PR and create a job for this detector, we can leave it failing until  this merges, but this way we'll have quick feedback

https://github.com/kubernetes/kubernetes/pull/95400#issuecomment-1071624107

> @aojea should that PR be feature gated?

no need for feature gate, is not a new API, is just implementing one of the local detectors that was approved here, is opt-in 

https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules/README.md#iptables---masquerade-off-cluster-traffic-to-services-by-node-ip

wedaly commented:
> There is a job that will run periodically using this detector, there is no more room for presubmits :/ , but this change is not very risky and @wedaly tested it. By the way, @wedaly can you please submit a PR and create a job for this detector, we can leave it failing until this merges, but this way we'll have quick feedback
> 

Sure thing! https://github.com/kubernetes/test-infra/pull/25743

khenidak commented:
Thanks @aojea @wedaly 

/lgtm

aojea commented:
@danwinship unhold when you are happy
/approve

aojea commented:
it needs Tim approval oo
/assign @thockin

thockin commented on pkg/proxy/apis/config/types.go:
```diff
@@ -107,6 +107,18 @@ type KubeProxyWinkernelConfiguration struct {
+	// BridgeInterfaceName is a string argument which represents a single bridge interface name.
+	// Kube-proxy considers traffic as local if originating from this given bridge.
+	// This argument should be set if DetectLocalMode is set to BridgeInterfaceName.
+	BridgeInterfaceName string
```
This is super nitpicky, but for naming I think I'd go "BridgeInterface" and "InterfaceNamePrefix" (both field names and consts for DetectLocalMode)

tssurya replied:
ack let me change this.

tssurya replied:
done.

thockin commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -122,6 +122,9 @@ oomScoreAdj: 17
 udpIdleTimeout: 123ms
 detectLocalMode: "ClusterCIDR"
+detectLocal:
+  bridgeInterfaceName: "eth0"
```
nit: s/eth0/cbr0

tssurya replied:
done.

thockin commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -317,3 +323,11 @@ func validateShowHiddenMetricsVersion(version string, fldPath *field.Path) field
+
+func validateInterface(iface string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(iface) <= 0 {
```
len() can not be < 0

tssurya replied:
done.

thockin commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -87,3 +87,62 @@ func (d *detectLocalByCIDR) IfLocal() []string {
+// NewDetectLocalByBridgeInterfaceName implements the LocalTrafficDetector interface using a bridge interface name.
+// This can be used when a bridge can be used to capture the notion of local traffic from pods.
+func NewDetectLocalByBridgeInterfaceName(interfaceName string) (LocalTrafficDetector, error) {
+	if len(interfaceName) <= 0 {
```
len() can not be < 0

tssurya replied:
yikes, sorry that was stupid of me.

thockin commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -87,3 +87,62 @@ func (d *detectLocalByCIDR) IfLocal() []string {
+// This can be used when a pod interface prefix can be used to capture the notion of local traffic. Note
+// that this will match on all interfaces that start with the given prefix.
+func NewDetectLocalByInterfacePrefix(interfacePrefix string) (LocalTrafficDetector, error) {
+	if len(interfacePrefix) <= 0 {
```
len() == 0

tssurya replied:
done.

thockin reviewed (commented):
I'm not super concerned about not having e2e, but shouldn't we have a proxy level unit test to cover this (to show a full, targetted dump)

tssurya commented:
Thanks Tim for reviewing!

> I'm not super concerned about not having e2e, but shouldn't we have a proxy level unit test to cover this (to show a full, targetted dump)

I did have full blown ones that show all the rules that get created in iptables_test.go but then I got a review comment saying since we already have coverage for the iptable rule that gets created in the traffic_test.go plus the common tests show all the rules I'd be adding redundant tests: https://github.com/kubernetes/kubernetes/pull/95400#discussion_r799061610 so I removed them.

thockin commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -87,3 +87,62 @@ func (d *detectLocalByCIDR) IfLocal() []string {
+		return nil, fmt.Errorf("no interface prefix set")
+	}
+	return &detectLocalByInterfacePrefix{
+		ifLocal:    []string{"-i", interfacePrefix + string('+')},
```
This casting of rune to string is unusual and raises suspicion - why not `interfacePrefix + "+"` ?

tssurya replied:
done.

thockin commented:
ping here when you push a new version

tssurya commented:
waiting for CI to be green...

thockin reviewed (commented):
Thanks!

/lgtm
/approve

tssurya commented:
/remove hold

tssurya commented:
@khenidak : are you ok with removing the hold?

tssurya commented:
/hold cancel
based on the lgtms

tssurya commented:
/test pull-kubernetes-e2e-kind-ipv6

tssurya commented:
/test pull-kubernetes-integration

cici37 commented:
/triage accepted
